services/system/Transact/src/RTransact.cpp

```diff
@@ -177,129 +179,327 @@ namespace
       }
    }
-   if (json)
+
+   void stopTracking(const std::vector<psibase::Checksum256>& txids)
    {
-      JsonHttpReply<TransactionTraceRef&> reply{.contentType = "application/json", .body{pruned}};
-      ActionViewBuilder<HttpServer>       http{getReceiver(), HttpServer::service};
-      auto                                action = http.sendReply(0, reply);
+      auto pendingTxTable = RTransact::Subjective{}.open<PendingTransactionTable>();
+      auto dataTable      = RTransact::Subjective{}.open<TransactionDataTable>();
 
-      for (auto client : row->clients)
-         if (client.json)
+      PSIBASE_SUBJECTIVE_TX
+      {
+         for (auto id : txids)
          {
-            psio::get<0>(action->rawData().value()) = client.socket;
-            call(action.data(), action.size());
+            pendingTxTable.erase(id);
+            dataTable.erase(id);
          }
+      }
    }
-   if (bin)
+
+   void sendReply(const psibase::Checksum256&                 id,
+                  psio::view<const psibase::TransactionTrace> trace,
+                  const ClientFilter&                         clientFilter = noFilter)
    {
-      FracpackHttpReply<TransactionTraceRef&> reply{.contentType = "application/octet-stream",
-                                                    .body{pruned}};
-      ActionViewBuilder<HttpServer>           http{getReceiver(), HttpServer::service};
-      auto                                    action = http.sendReply(0, reply);
+      TransactionTraceRef pruned     = PruneTrace{true}(trace);
+      auto [jsonClients, binClients] = claimClientReply(id, clientFilter);
+
+      if (jsonClients.empty() && binClients.empty())
+         return;
 
-      for (auto client : row->clients)
-         if (!client.json)
+      ActionViewBuilder<HttpServer> http{getReceiver(), HttpServer::service};
+
+      if (!jsonClients.empty())
+      {
+         JsonHttpReply<TransactionTraceRef&> reply{.contentType = "application/json",
+                                                   .body{pruned}};
+         auto                                action = http.sendReply(0, reply);
+
+         for (auto socket : jsonClients)
          {
-            psio::get<0>(action->rawData().value()) = client.socket;
+            psio::get<0>(action->rawData().value()) = socket;
             call(action.data(), action.size());
          }
+      }
+
+      if (!binClients.empty())
+      {
+         FracpackHttpReply<TransactionTraceRef&> reply{.contentType = "application/octet-stream",
+                                                       .body{pruned}};
+         auto                                    action = http.sendReply(0, reply);
+
+         for (auto socket : binClients)
+         {
+            psio::get<0>(action->rawData().value()) = socket;
+            call(action.data(), action.size());
+         }
+      }
    }
+
+   void sendReplies(const std::vector<psibase::Checksum256>& txids)
+   {
+      stopTracking(txids);
```

Comment:
If the node is killed after this, it can leave orphaned traces behind.

Reply:
Not sure how to handle this. I've moved the deletion to the end of `sendReplies`, but if I understand correctly, this is still a risk.

Reply:
The order is now:
1. Remove trace
2. Remove client
3. Remove from queue

I think this is safe, as the queue entries drive cleanup. After a node restart, we don't care about the client and trace, except for the making sure that they get cleaned up.